core/src/main/java/org/apache/brooklyn/core/typereg/AbstractTypePlanTransformer.java

```diff
@@ -0,0 +1,130 @@
+
+    @Override
+    public void injectManagementContext(ManagementContext mgmt) {
+        this.mgmt = mgmt;
+    }
+
+    private final String format;
+    private final String formatName;
+    private final String formatDescription;
+    
+    protected AbstractTypePlanTransformer(String format, String formatName, String formatDescription) {
+        this.format = format;
+        this.formatName = formatName;
+        this.formatDescription = formatDescription;
+    }
+    
+    @Override
+    public String getFormatCode() {
+        return format;
+    }
+
+    @Override
+    public String getFormatName() {
+        return formatName;
+    }
+
+    @Override
+    public String getFormatDescription() {
+        return formatDescription;
+    }
+
+    @Override
+    public String toString() {
+        return getFormatCode()+":"+JavaClassNames.simpleClassName(this);
+    }
+    
+    @Override
+    public double scoreForType(RegisteredType type, RegisteredTypeLoadingContext context) {
+        if (getFormatCode().equals(type.getPlan().getPlanFormat())) return 1;
+        if (type.getPlan().getPlanFormat()==null)
+            return scoreForNullFormat(type.getPlan().getPlanData(), type, context);
+        else
+            return scoreForNonmatchingNonnullFormat(type.getPlan().getPlanFormat(), type.getPlan().getPlanData(), type, context);
+    }
+
+    protected abstract double scoreForNullFormat(Object planData, RegisteredType type, RegisteredTypeLoadingContext context);
+    protected abstract double scoreForNonmatchingNonnullFormat(String planFormat, Object planData, RegisteredType type, RegisteredTypeLoadingContext context);
+
+    /** delegates to more specific abstract create methods,
+     * and performs common validation and customisation of the items created.
+     * <p>
+     * this includes:
+     * <li> setting the {@link AbstractBrooklynObjectSpec#catalogItemId(String)}
+     */
+    @Override
+    public Object create(final RegisteredType type, final RegisteredTypeLoadingContext context) {
+        try {
+            return validate(new RegisteredTypeKindVisitor<Object>() {
+                @Override protected Object visitSpec(RegisteredType type) {
+                    try { 
+                        AbstractBrooklynObjectSpec<?, ?> result = createSpec(type, context);
+                        result.catalogItemId(type.getId());
+                        return result;
+                    } catch (Exception e) { throw Exceptions.propagate(e); }
+                }
+                @Override protected Object visitBean(RegisteredType type) {
+                    try { 
+                        return createBean(type, context);
+                    } catch (Exception e) { throw Exceptions.propagate(e); }
+                }
+                
+            }.visit(type), type, context);
+        } catch (UnsupportedTypePlanException e) {
+            // no logging
+            throw Exceptions.propagate(e);
```

Comment:
My understanding is that this is just to prevent the next catch from taking over - worth a comment. May be even throwing the same exception instead of calling in to `Exceptions.propagate` to make it more visible (even if the end result is the same).